src/libraries/System.Console/src/System/IO/KeyParser.cs

```diff
@@ -15,6 +15,38 @@ internal static class KeyParser
     private const int MinimalSequenceLength = 3;
     private const int SequencePrefixLength = 2; // ^[[ ("^[" stands for Escape)
 
+    // if provided buffer contains Bracketed Paste Start or End tag, the buffer is sliced to pasted content only
+    // In theory, a single buffer can contain end tag from previous paste followed immediately by start tag from next.
+    internal static ReadOnlySpan<byte> RemoveBracketedPasteTags(ReadOnlySpan<byte> receivedBytes)
+    {
+        // The Terminal is configured to return from read() as soon as at least byte is available for reading.
+        // This is achieved by setting Termios.c_cc[VMIN] = 1 in the native initialization logic.
+        // So in the most common case, when bracketed paste is not enabled and the user presses a single key,
+        // the receivedBytes contains a single byte for simple ASCII chars and just a few bytes for escape sequences.
+        // So the expensive search performed below is not common.
+        const int TagLength = 6; // "\u001B[200~".Length
+        while (receivedBytes.Length >= TagLength)
+        {
+            if (IsTag(receivedBytes.Slice(0, TagLength)))
+            {
+                receivedBytes = receivedBytes.Slice(TagLength);
+            }
+            else if (IsTag(receivedBytes.Slice(receivedBytes.Length - TagLength)))
+            {
+                receivedBytes = receivedBytes.Slice(0, receivedBytes.Length - TagLength);
+            }
+            else
+            {
+                break;
+            }
+        }
+
+        return receivedBytes;
+
+        static bool IsTag(ReadOnlySpan<byte> b)
+            => b[0] == '\u001B' && b[1] == '[' && b[2] == '2' && b[3] == '0' && (b[4] == '0' || b[4] == '1') && b[5] == '~';
```

Comment:
I may be reading this wrong, but it looks like this will also pass malformed sequences where the end tag comes before the start tag -- a frame shift could easily result in this happening "in the wild", right? If we miss part of the introducer, we might detect a span flowing from the next finalizer to the next introducer as having been a bracketed paste.

Reply:
> end tag comes before the start tag

As I wrote in the unit tests comments, this handles hypothetical scenario where `read()` returns end tag from previous pasted content followed by start tag from next pasted content.

Example: user pastes "AB", then "CD". 

1st read returns: `<startTag>AB`
2nd read returns: `</endTag><startTag>CD</endTag>`

https://github.com/dotnet/runtime/blob/da4d3511b1a1e6021ea07ea1e8a148fb499da8c7/src/libraries/System.Console/tests/KeyParserTests.cs#L413-L416

I could stick only to the test cases I've confirmed with manual testing, but each Terminal may implement this feature differently and I can't guarantee that we are going to handle all of them properly.

Reply:
Should we validate that we get at least one start and one end tag (and ignore any nested tags that may occur)?